### PR TITLE
License forbidden

### PR DESCRIPTION
**Pull request by lionellloh**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
This PR allows Homebrew users to specify a collection of "Forbidden Licenses" as an environment variable which will prevent them from installing a package if it or its dependency package has a forbidden license. 

`export HOMEBREW_FORBIDDEN_LICENSES="0BSD MIT AML"` 
`ENV[HOMEBREW_FORBIDDEN_LICENSES]` is parsed as a space delimited string into a list of the following variables` ["0BSD", "MIT", "AML" ]`

When `brew install xxx` is run, we will check if xxx or any of its dependency uses a license that belongs to the list `["0BSD", "MIT", "AML"]`

### CODE REVIEW

File: Library/Homebrew/cmd/install.rb [hunk 1]
```diff
@@ -258,6 +258,7 @@ def install
 
     formulae.each do |f|
       Migrator.migrate_if_needed(f)
+      forbidden_license_check(f) unless ENV["HOMEBREW_FORBIDDEN_LICENSES"].blank?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/install.rb:261]: Probably want this in `formula_installer.rb` somewhere so it's handled by `upgrade` and `reinstall`, too. Either `def prelude` or `def check_install_sanity`?

File: Library/Homebrew/cmd/install.rb [hunk 2]
```diff
@@ -338,3 +339,23 @@ def install_formula(f)
     ofail e.message
   end
 end
+
+def forbidden_license_check(f)
+  forbidden_licenses = ENV["HOMEBREW_FORBIDDEN_LICENSES"].split(" ")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/install.rb:344]:
```suggestion
  forbidden_licenses = Homebrew::EnvConfig.forbidden_licenses.split(" ")
```

or could even add a custom method there to make it return an Array?

File: Library/Homebrew/test/cmd/install_spec.rb [hunk 3]
```diff
@@ -17,6 +17,26 @@
     expect(HOMEBREW_CELLAR/"testball1/0.1/foo/test").not_to be_a_file
   end
 
+  it "does not install formulae with forbidden license" do
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/cmd/install_spec.rb:20]: Sorry to waste your time on these but we want to avoid adding more integration tests here (because they are really slow). It may be that there's no valid tests for this.

File: .github/codecov.yml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+fixes:
```
Comment (MikeMcQuaid) [inline .github/codecov.yml:1]: oops, this shouldn't be here. Might need to `git fetch --all; git rebase --interactive origin/master; git push --force-with-lease` and ensure only the relevant commits are in here.

File: Library/Homebrew/formula_installer.rb [hunk 5]
```diff
@@ -149,6 +149,8 @@ def install_bottle_for?(dep, build)
   def prelude
     Tab.clear_cache
     verify_deps_exist unless ignore_deps?
+    Homebrew.forbidden_license_check(formula) unless ENV["HOMEBREW_FORBIDDEN_LICENSES"].blank?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb:153]: Could maybe call this unconditionally and just make it a no-op if `ENV["HOMEBREW_FORBIDDEN_LICENSES"].blank?`

File: docs/Manpage.md [hunk 6]
```diff
@@ -1394,6 +1394,9 @@ Note that environment variables must have a value set to be detected. For exampl
 
     *Default:* `15`.
 
+  * `HOMEBREW_FORBIDDEN_LICENSES`:
+    Use this environment variable to define a blacklist of space separated licenses and Homebrew will avoid installing the packages with those licenses.
```
Comment (MikeMcQuaid) [inline docs/Manpage.md:1398]:
```suggestion
    Use this environment variable to define a denylist of space separated licenses and Homebrew will refuse to install packages known to have those licenses.
```

File: Library/Homebrew/test/cmd/install_spec.rb (hunk 3 above)
Comment (lionellloh) [reply Library/Homebrew/test/cmd/install_spec.rb:20]: no worries

File: .github/codecov.yml (hunk 4 above)
Comment (lionellloh) [reply .github/codecov.yml:1]: Hi @MikeMcQuaid , I got a little confused about why there are so many commits appearing here as well. What I did was `git merge master` on my feature branch which I thought was a normal practice before we merge into master.

Comment (MikeMcQuaid) [reply .github/codecov.yml:1]: @lionellloh Probably best to stick with `git merge origin/master` to ensure there's nothing extra on your `master`. In this case I'd `rebase`; it's a little more obvious locally what'll show up in the diff view.

Comment (lionellloh) [reply .github/codecov.yml:1]: Ah ok, I think I understand what went wrong.

**Comment by lionellloh** [comment]
@MikeMcQuaid Your suggestion worked like a charm! There's so much more I need to learn about git...

File: Library/Homebrew/cmd/install.rb [hunk 7]
```diff
@@ -259,6 +259,7 @@ def install
 
     formulae.each do |f|
       Migrator.migrate_if_needed(f)
+      forbidden_license_check(f) unless ENV["HOMEBREW_FORBIDDEN_LICENSES"].blank?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/install.rb:262]: Think this can be removed in favour of the `formula_installer` check.

File: Library/Homebrew/cmd/install.rb [hunk 8]
```diff
@@ -341,4 +342,28 @@ def install_formula(f)
...
+    fi = FormulaInstaller.new(f)
+    fi.compute_dependencies.each do |dep, _|
+      dep_f = dep.to_formula
+      next unless forbidden_licenses.include? dep_f.license
+
+      raise CannotInstallFormulaError, <<~EOS
+        The installation of #{f.name} has a dependency on #{dep.name} with a forbidden license #{dep_f.license}.
+      EOS
+    end
+  end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/install.rb:368]: I'd consider moving all of this to `formula_installer.rb` given it's having to make use of it anyway.

File: Library/Homebrew/env_config.rb [hunk 9]
```diff
@@ -111,6 +111,10 @@ module EnvConfig
         description: "Output this many lines of output on formula `system` failures.",
         default:     15,
       },
+      HOMEBREW_FORBIDDEN_LICENSES:        {
+        description: "Use this environment variable to define a blacklist of space separated licenses and Homebrew " \
```
Comment (MikeMcQuaid) [inline Library/Homebrew/env_config.rb:115]:
```suggestion
        description: "Use this environment variable to define a denylist of space separated licenses and Homebrew " \
```

File: Library/Homebrew/formula_installer.rb [hunk 10]
```diff
@@ -1102,4 +1105,30 @@ def puts_requirement_messages
...
   end
+
+  def env_forbidden_licenses
+    Homebrew::EnvConfig.forbidden_licenses.split(" ")
+  end
+
+  def forbidden_license_check(f)
+    return if ENV["HOMEBREW_FORBIDDEN_LICENSES"].blank?
+
+    forbidden_licenses = env_forbidden_licenses
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb:1116]:
```suggestion
  def forbidden_license_check
    forbidden_licenses = Homebrew::EnvConfig.forbidden_licenses.split(" ")
    return if forbidden_licenses.blank?
```

File: Library/Homebrew/formula_installer.rb [hunk 11]
```diff
@@ -1102,4 +1105,30 @@ def puts_requirement_messages
...
+
+    if forbidden_licenses.include? f.license
+      raise CannotInstallFormulaError, <<~EOS
+        #{f.name} has a forbidden license #{f.license}.
+      EOS
+    end
+
+    fi = FormulaInstaller.new(f)
+    fi.compute_dependencies.each do |dep, _|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb:1125]:
```suggestion
    compute_dependencies.each do |dep, _|
```

File: Library/Homebrew/formula_installer.rb [hunk 12]
```diff
@@ -1102,4 +1105,30 @@ def puts_requirement_messages
...
+  def forbidden_license_check(f)
+    return if ENV["HOMEBREW_FORBIDDEN_LICENSES"].blank?
+
+    forbidden_licenses = env_forbidden_licenses
+
+    if forbidden_licenses.include? f.license
+      raise CannotInstallFormulaError, <<~EOS
+        #{f.name} has a forbidden license #{f.license}.
+      EOS
+    end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb:1122]:
```suggestion
    if forbidden_licenses.include? formula.license
      raise CannotInstallFormulaError, <<~EOS
        #{formula.name} has a forbidden license #{formula.license}.
      EOS
    end
```

File: Library/Homebrew/formula_installer.rb [hunk 13]
```diff
@@ -1102,4 +1105,24 @@ def puts_requirement_messages
 
     $stderr.puts @requirement_messages
   end
+
+  def forbidden_license_check
+    forbidden_licenses = (Homebrew::EnvConfig.forbidden_licenses || "").split(" ")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb:1110]:
```suggestion
    forbidden_licenses = Homebrew::EnvConfig.forbidden_licenses.to_s.split(" ")
```

File: Library/Homebrew/env_config.rb [hunk 14]
```diff
@@ -111,6 +111,10 @@ module EnvConfig
         description: "Output this many lines of output on formula `system` failures.",
         default:     15,
       },
+      HOMEBREW_FORBIDDEN_LICENSES:        {
+        description: "Use this environment variable to define a denylist of space separated licenses and Homebrew " \
+                     "will avoid installing the packages with those licenses.",
```
Comment (jonchang) [inline Library/Homebrew/env_config.rb:116]: Maybe reword this to match the other descriptions on the page (e.g. none of them explicitly refer to "environment variables"). I suggest...

> A space-separated list of licenses. Homebrew will refuse to install a formula if that formula or any of its dependencies has a license on this list.

**Comment by lionellloh** [comment]
Hi @MikeMcQuaid, here are the tests I ran: 

1. Test that Package with forbidden license cannot be installed.
2. Test that Package with license that is not forbidden can be installed. 
3. Test that Package without license can be installed 

```
export HOMEBREW_FORBIDDEN_LICENSES="Apache-2.0"   
brew install akamai # Error: akamai has a forbidden license Apache-2.0
brew install cask #success as cask's license is GPL-3.0
brew install aggregate #success as aggregate has no license 
```

1. Test that space separated string is parsed properly
2. Test that forbidden license works for `brew reinstall` too 
```
export HOMEBREW_FORBIDDEN_LICENSES="Apache-2.0 GPL-3.0"   
brew install akamai # Error: akamai has a forbidden license Apache-2.0
brew reinstall cask #fail as cask's license is GPL-3.0
```

**Comment by MikeMcQuaid** [comment]
Thanks again @lionellloh, great work as usual!
